RootInteractive/InteractiveDrawing/bokeh/surface3d.ts

```diff
@@ -0,0 +1,164 @@
+import {LayoutItem} from "core/layout"
+import * as p from "core/properties"
+
+declare namespace vis {
+  class Graph3d {
+    constructor(el: HTMLElement, data: object, OPTIONS: object)
+    setData(data: vis.DataSet): void
+    setOptions(options: object): void
+  }
+
+  class DataSet {
+    add(data: unknown): void
+  }
+}
+
+// This defines some default options for the Graph3d feature of vis.js
+// See: http://visjs.org/graph3d_examples.html for more details.
+const OPTIONS = {
+  width: '400px',
+  height: '400px',
+  // style: 'surface',
+  style: 'dot-color',
+  showPerspective: true,
+  showGrid: true,
+  keepAspectRatio: true,
+  verticalRatio: 1.0,
+  legendLabel: 'legendLabel',
+  cameraPosition: {
+    horizontal: -0.35,
+    vertical: 0.22,
+    distance: 1.8,
+  },
+}
+// To create custom model extensions that will render on to the HTML canvas
+// or into the DOM, we must create a View subclass for the model.
+//
+// In this case we will subclass from the existing BokehJS ``LayoutDOMView``
+export class Surface3dView extends LayoutDOMView {
+  model: Surface3d
+
+  private _graph: vis.Graph3d
+
+  initialize(): void {
+    super.initialize()
+
+    const url = "https://cdnjs.cloudflare.com/ajax/libs/vis/4.16.1/vis.min.js"
+    const script = document.createElement("script")
+    script.onload = () => this._init()
+    script.async = false
+    script.src = url
+    document.head.appendChild(script)
+  }
+
+  private _init(): void {
+    // Create a new Graph3s using the vis.js API. This assumes the vis.js has
+    // already been loaded (e.g. in a custom app template). In the future Bokeh
+    // models will be able to specify and load external scripts automatically.
+    //
+    // BokehJS Views create <div> elements by default, accessible as this.el.
+    // Many Bokeh views ignore this default <div>, and instead do things like
+    // draw to the HTML canvas. In this case though, we use the <div> to attach
+    // a Graph3d to the DOM.
+    this._graph = new vis.Graph3d(this.el, this.get_data(), OPTIONS)
+    if(this.model.options3D !== null){
+      this._graph.setOptions(this.model.options3D)
+    }
+
+    // Set a listener so that when the Bokeh data source has a change
+    // event, we can process the new data
+    this.connect(this.model.data_source.change, () => {
+      this._graph.setData(this.get_data())
+      this._graph.setOptions(this.model.options3D)
+    })
+
+  }
+
+  // This is the callback executed when the Bokeh data has an change. Its basic
+  // function is to adapt the Bokeh data source to the vis.js DataSet format.
+  get_data(): vis.DataSet {
+    const data = new vis.DataSet()
+    const source = this.model.data_source
+    for (let i = 0; i < source.get_length()!; i++) {
+      data.add({
+        x: source.data[this.model.x][i],
+        y: source.data[this.model.y][i],
+        z: source.data[this.model.z][i],
+        style: source.data[this.model.style][i],
+      })
+    }
+    return data
+  }
+
```

Comment:
get_option callback function which we discussed yesterday is good idea
e.g to witch from graph-> to surface via  drawing switch
similar as we discussed for the legend

Reply:
For later